File: src/ui/icons.rs
//! Solar icon set (https://solar-icons.vercel.app) — встроенные SVG, красятся через tint()
use egui::{Color32, Ui};

macro_rules! icon_set {
    ($($key:ident => $file:literal),* $(,)?) => {
        pub enum Icon { $($key),* }

        fn bytes(icon: &Icon) -> &'static [u8] {
            match icon {
                $(Icon::$key => include_bytes!(concat!("icons/svg/", $file, ".svg"))),*
            }
        }

        fn uri(icon: &Icon) -> &'static str {
            match icon {
                $(Icon::$key => concat!("bytes://solar/", $file, ".svg")),*
            }
        }
    };
}

icon_set! {
    Folder => "folder-bold",
    FolderWithFiles => "folder-with-files-linear",
    AddCircleLinear => "add-circle-linear",
    AddCircleBold => "add-circle-bold",
    AddSquare => "add-square-linear",
    Upload => "upload-linear",
    Download => "download-linear",
    DownloadSquare => "download-square-bold",
    Clock => "clock-circle-linear",
    Refresh => "refresh-linear",
    Trash => "trash-bin-minimalistic-linear",
    Pen => "pen-linear",
    History => "history-linear",
    ServerSquare => "server-square-linear",
    Home => "home-2-bold",
    Monitor => "monitor-bold",
    Document => "document-bold",
    DocumentText => "document-text-bold",
    Ssd => "ssd-square-linear",
    Database => "database-linear",
    DangerTriangle => "danger-triangle-bold",
    LockPassword => "lock-password-bold",
    ArrowDown => "alt-arrow-down-linear",
    ArrowUp => "alt-arrow-up-linear",
    CheckCircle => "check-circle-bold",
    PlayCircle => "play-circle-linear",
    PauseCircle => "pause-circle-linear",
    FileText => "file-text-bold",
    Gallery => "gallery-bold",
    Code => "code-bold",
    Archive => "archive-down-minimlistic-bold",
    ShieldKeyhole => "shield-keyhole-bold",
    File => "file-bold",
    Star => "star-bold",
    Videocamera => "videocamera-record-bold",
    MusicNote => "music-note-3-bold",
    Link => "link-bold",
    CloseCircle => "close-circle-linear",
    Settings => "settings-linear",
}

/// Рисует иконку заданного размера и цвета (tint) в текущей позиции ui.
pub fn icon(ui: &mut Ui, icon: Icon, size: f32, color: Color32) -> egui::Response {
    let image = egui::Image::from_bytes(uri(&icon), bytes(&icon))
        .tint(color)
        .fit_to_exact_size(egui::vec2(size, size));
    ui.add(image)
}

/// Тот же рендер, но для использования внутри builder-паттернов (кнопки и т.д.) — возвращает Image.
pub fn image(icon: Icon, size: f32, color: Color32) -> egui::Image<'static> {
    egui::Image::from_bytes(uri(&icon), bytes(&icon))
        .tint(color)
        .fit_to_exact_size(egui::vec2(size, size))
}

/// Цвет + иконка по расширению файла — палитра Clay
pub fn file_icon_for(ext: &str) -> (Icon, Color32) {
    use crate::ui::theme::*;
    match ext {
        "pdf" => (Icon::FileText, RED),
        "xlsx" | "xls" | "doc" | "docx" => (Icon::Document, GREEN),
        "png" | "jpg" | "jpeg" | "gif" | "webp" | "svg" => (Icon::Gallery, ACCENT),
        "mp4" | "mkv" | "avi" | "mov" => (Icon::Videocamera, ACCENT),
        "mp3" | "ogg" | "flac" | "wav" => (Icon::MusicNote, GREEN),
        "zip" | "tar" | "gz" | "bz2" | "xz" | "7z" => (Icon::Archive, TEXT_DIM),
        "yaml" | "yml" | "toml" | "json" => (Icon::Code, YELLOW),
        "php" | "java" | "class" | "jar" => (Icon::Code, ACCENT_DIM),
        "rs" | "py" | "sh" | "bash" | "fish" | "zsh" => (Icon::Code, GREEN),
        "html" | "css" | "js" | "ts" => (Icon::Code, ACCENT),
        "log" => (Icon::DocumentText, TEXT_HINT),
        "htaccess" => (Icon::ShieldKeyhole, TEXT_DIM),
        "txt" | "md" => (Icon::DocumentText, TEXT_DIM),
        _ => (Icon::File, TEXT_DIM),
    }
}

File: frontend/js/icons.js
// Thin rendering helper around the Solar BoldDuotone icon paths from
// icons-data.js. Every icon has two path layers: a full-opacity "detail"
// layer and a 50%-opacity "soft" background layer.
import { ICONS } from "./icons-data.js";

const softLayerCache = new Map();

function withSolidSecondaryLayer(body) {
  const cached = softLayerCache.get(body);
  if (cached) return cached;
  const processed = body.replace(/<path([^>]*)\/>/g, (match, attrs) => {
    if (!/\sopacity="/.test(attrs)) return match;
    const withoutOpacity = attrs.replace(/\s*opacity="[^"]*"/, "");
    return `<path${withoutOpacity} class="icon-soft"/>`;
  });
  softLayerCache.set(body, processed);
  return processed;
}

export function iconMarkup(name, size = 16) {
  const body = ICONS[name];
  if (!body) return "";
  return `<svg class="icon" viewBox="0 0 24 24" width="${size}" height="${size}" aria-hidden="true">${withSolidSecondaryLayer(body)}</svg>`;
}

export function iconEl(name, size = 16) {
  const wrap = document.createElement("span");
  wrap.style.display = "inline-flex";
  wrap.innerHTML = iconMarkup(name, size);
  return wrap.firstElementChild;
}

// ── File-type icon lookup ──────────────────────────────────────────────

const FILE_ICONS = {
  pdf: ["fileText", "icon--important"],
  xlsx: ["document", "icon--important"],
  xls: ["document", "icon--important"],
  doc: ["document", "icon--important"],
  docx: ["document", "icon--important"],
  png: ["gallery", "icon--important"],
  jpg: ["gallery", "icon--important"],
  jpeg: ["gallery", "icon--important"],
  gif: ["gallery", "icon--important"],
  webp: ["gallery", "icon--important"],
  svg: ["gallery", "icon--important"],
  mp4: ["videocamera", "icon--important"],
  mkv: ["videocamera", "icon--important"],
  avi: ["videocamera", "icon--important"],
  mov: ["videocamera", "icon--important"],
  webm: ["videocamera", "icon--important"],
  mp3: ["musicNote", "icon--important"],
  ogg: ["musicNote", "icon--important"],
  flac: ["musicNote", "icon--important"],
  wav: ["musicNote", "icon--important"],
  m4a: ["musicNote", "icon--important"],
  zip: ["archive", "icon--important"],
  tar: ["archive", "icon--important"],
  gz: ["archive", "icon--important"],
  bz2: ["archive", "icon--important"],
  xz: ["archive", "icon--important"],
  "7z": ["archive", "icon--important"],
  yaml: ["code", "icon--important"],
  yml: ["code", "icon--important"],
  toml: ["code", "icon--important"],
  json: ["code", "icon--important"],
  php: ["code", "icon--important"],
  java: ["code", "icon--important"],
  class: ["code", "icon--important"],
  jar: ["code", "icon--important"],
  rs: ["code", "icon--important"],
  py: ["code", "icon--important"],
  sh: ["code", "icon--important"],
  bash: ["code", "icon--important"],
  fish: ["code", "icon--important"],
  zsh: ["code", "icon--important"],
  html: ["code", "icon--important"],
  css: ["code", "icon--important"],
  js: ["code", "icon--important"],
  ts: ["code", "icon--important"],
  sql: ["database", "icon--important"],
  db: ["database", "icon--important"],
  sqlite: ["database", "icon--important"],
  env: ["shieldKeyhole", "icon--important"],
  pem: ["key", "icon--important"],
  pub: ["key", "icon--important"],
  log: ["documentText", "icon--inactive"],
  htaccess: ["shieldKeyhole", "icon--inactive"],
  conf: ["settings", "icon--inactive"],
  ini: ["settings", "icon--inactive"],
  txt: ["documentText", "icon--inactive"],
  md: ["documentText", "icon--inactive"],
};

export function fileIconFor(name) {
  const ext = name.includes(".") ? name.split(".").pop().toLowerCase() : "";
  return FILE_ICONS[ext] ?? ["file", "text-dim"];
}

// ── Quick-access sidebar icon map ──────────────────────────────────────

export const QUICK_ACCESS_ICONS = {
  Home: "home",
  Desktop: "monitor",
  Documents: "document",
  Downloads: "downloadSquare",
  Pictures: "gallery",
  Music: "musicNote",
  Videos: "videocamera",
};
